ssl/d1_lib.c

```diff
@@ -75,29 +75,19 @@ int dtls1_new(SSL *ssl)
 
     if (!ssl3_new(ssl))
         return 0;
-    if ((d1 = OPENSSL_zalloc(sizeof(*d1))) == NULL) {
+    if ((d1 = OPENSSL_zalloc(sizeof(*d1) + 2 * sizeof(struct pqueue_st))) == NULL) {
```

Comment:
You cannot do it like this. The pqueue_st structures would be misaligned. What you could do is to make the structures members of DTLS1_STATE.

Reply:
Can you explain why this goes wrong? Not that I don't believe you, I just want to learn :)

Reply:
Any larger-than-1-byte primitive data type (int, pointer) must be placed into memory at aligned addresses. Although X86 processors fully handle the unaligned access, other architectures might be different - i.e. they won't be able to handle it and it will cause exceptions. Even on architectures which are able to handle unaligned accesses the access might be slower than in case of aligned access.

Unless the structure contains only 1byte sized members (or arrays of such) the structure must be placed at aligned addresses. (Maybe this is true even for structures with 1 byte sized members, but I am not sure about that.)

Reply:
Thanks!